### PR TITLE
DOCKER-806: 'docker ps -a' shows an empty string 'STATUS' for containers that stopped due to a CN crash/hard-stop

### PR DESCRIPTION
**Pull request by trentm**

### CODE REVIEW

File: lib/backends/sdc/utils.js [hunk 1]
```diff
@@ -163,6 +163,68 @@ function getPublishedPorts(opts, fwrules, cb) {
...
+             *
+             * "Exited" isn't really correct here if it never started, so let's
+             * not lie and just say "Stopped". We are already going to differ
+             * from Docker-docker anyway.
+             */
+            status = 'Stopped';
+        }
+    } else {
+        // TODO: handle other states here! See "VM STATES" in `man vmadm`.
+        status = '';
```
Comment (joshwilsdon) [inline lib/backends/sdc/utils.js]: I think Docker now shows these as 'Created'? ```root@ubuntu:~# docker ps -a
CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS                      PORTS               NAMES
84be3a553d59        nginx               "nginx -g 'daemon off"   5 weeks ago         Up 12 seconds               80/tcp, 443/tcp     berserk_sammet
1f1b2be7c0cc        alpine              "/bin/sh"                3 months ago        Exited (0) 3 months ago                         evil_einstein
09c8a7078b94        ubuntu              "/bin/bash"              3 months ago        Exited (0) 3 months ago                         stupefied_mestorf
05bf9cb6dee1        ubuntu              "/bin/bash"              3 months ago        Exited (130) 3 months ago                       condescending_nobel
b4efcebe29fd        busybox             "/bin/sh"                3 months ago        Created                                         stoic_hopper```

File: lib/backends/sdc/utils.js [hunk 2]
```diff
@@ -163,6 +163,68 @@ function getPublishedPorts(opts, fwrules, cb) {
...
+            && vmobj.hasOwnProperty('exit_timestamp'))
+        {
+            exittime
+                = Math.floor((now - (new Date(vmobj.exit_timestamp))) / 1000);
+            status = 'Exited (' + vmobj.exit_status + ') '
+                + common.humanDuration(exittime) + ' ago';
+        } else {
+            /*
+             * Two known possibilities here:
+             * 1. The zone never started (e.g. if created, but not run; or
```
Comment (joshwilsdon) [inline lib/backends/sdc/utils.js]: I think this first case can be detected by the presence of docker:restartcount in the internal_metadata. If a VM has this property (even set to 0) it has been started at least once. When you do `docker create` this property is not added until the VM is started. So perhaps list as 'Created' when state==='stopped' and docker:restartcount doesn't exist?

File: lib/backends/sdc/utils.js [hunk 3]
```diff
@@ -163,6 +163,68 @@ function getPublishedPorts(opts, fwrules, cb) {
...
+             *    some failure before it got to running). Last sometime checked
```
Comment (joshwilsdon) [inline lib/backends/sdc/utils.js]: "Last sometime checked" seems like not what you were intending.

**Comment by trentm** [comment]
With joshw's suggestions:

```
$ docker --tls ps -a
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS                   PORTS               NAMES
74f3ab5b07b3        alpine              "sleep 3600"        5 minutes ago       Up 4 minutes                                 createthenstart0
870d3ca62d8f        alpine              "sleep 3600"        7 minutes ago       Created                                      created0
fd1121c20f68        alpine              "sleep 3600"        4 hours ago         Exited                                       again0
3d7bbe0019db        alpine              "sleep 3600"        5 hours ago         Exited (0) 4 hours ago                       another0
f249f3d93cb6        alpine              "sleep 3600"        6 hours ago         Exited (0) 6 hours ago                       running
0f82f1149d42        alpine              "sleep 3600"        26 hours ago        Exited                                       reverent_visvesvaraya
```

**Comment by trentm** [comment]
merged in commit 878e6c6c09e21d612df4f9639719a6ee65ef1e6d

File: lib/backends/sdc/utils.js [hunk 4]
```diff
@@ -188,32 +188,33 @@ function _containerStatusFromVmobj(opts) {
         uptime = Math.floor((now - (new Date(vmobj.create_timestamp))) / 1000);
         status = 'Provisioning ' + common.humanDuration(uptime);
     } else if (vmobj.state == 'stopped') {
+        /*
+         * Possibilities here:
+         * 1. We have `VM.exit_status` and `VM.exit_timestamp`. The VM exited
+         *    and zoneadmd saw it.
+         * 2. The zone never started. `vmadm` sets
+         *    `VM.internal_metadata['restartcount']` on first start.
+         * 2. The CN crashed or hard-stopped such that zoneadmd couldn't
```
Comment (joshwilsdon) [inline lib/backends/sdc/utils.js]: probably should be 3.

Comment (trentm) [inline lib/backends/sdc/utils.js]: heh

**Pull request closed without merging**